File: apps/api/src/routes/workspaces.ts
import { Hono } from 'hono'
import {
  createWorkspaceSchema,
  updateWorkspaceSchema,
  workspaceExportSchema,
} from '@clawbuddy/shared'
import type { WorkspaceExport } from '@clawbuddy/shared'
import { z } from 'zod'
import { workspaceService } from '../services/workspace.service.js'
import { capabilityService } from '../services/capability.service.js'
import { sandboxService } from '../services/sandbox.service.js'
import { settingsService } from '../services/settings.service.js'
import { channelService } from '../services/channel.service.js'
import { prisma } from '../lib/prisma.js'
import { decrypt } from '../services/crypto.service.js'
import { decryptConfigFields } from '../services/config-validation.service.js'
import { validateBody } from '../lib/validate.js'
import { buildResolvedRoleProviders } from '../lib/llm-resolver.js'
import { logger } from '../lib/logger.js'

const app = new Hono()

type ExportCapabilityRow = {
  slug: string
  configSchema: unknown
  config: unknown
}

type ExportChannelRow = {
  type: string
  name: string
  enabled: boolean
  config: unknown
}

const updateWorkspaceSettingsSchema = z.object({
  color: z.string().max(20).optional(),
  settings: z.record(z.unknown()).nullish(),
  autoExecute: z.boolean().optional(),
  permissions: z.object({ allow: z.array(z.string()) }).nullish(),
})

app.get('/', async (c) => {
  const workspaces = await workspaceService.list()
  return c.json({ success: true, data: workspaces })
})

app.post('/', async (c) => {
  const body = await c.req.json()
  const parsed = validateBody(createWorkspaceSchema, body)
  const workspace = await workspaceService.create(parsed)
  return c.json({ success: true, data: workspace }, 201)
})

// ── Workspace Import (must be before /:id to avoid param match) ──

app.post('/import', async (c) => {
  const body = await c.req.json()
  const parsed = validateBody(workspaceExportSchema, body)

  const skippedCapabilities: string[] = []
  const warnings: string[] = []

  // Create workspace
  const workspace = await workspaceService.create({
    name: `${parsed.workspace.name} (imported)`,
    description: parsed.workspace.description ?? undefined,
    color: parsed.workspace.color ?? undefined,
    settings: parsed.workspace.settings ?? undefined,
  })

  // Apply permissions and autoExecute
  await workspaceService.update(workspace.id, {
    permissions: parsed.workspace.permissions ?? undefined,
    autoExecute: parsed.workspace.autoExecute,
  })

  // Enable capabilities
  for (const cap of parsed.capabilities) {
    if (!cap.enabled) continue
    try {
      await capabilityService.enableCapability(
        workspace.id,
        cap.slug,
        cap.config as Record<string, unknown> | undefined,
      )
    } catch {
      skippedCapabilities.push(cap.slug)
    }
  }

  // Create channels (disabled by default)
  for (const ch of parsed.channels) {
    try {
      const config = ch.config as { botToken: string; [key: string]: unknown }
      if (!config.botToken) {
        warnings.push(`Channel "${ch.name}" skipped — no bot token`)
        continue
      }
      await channelService.create({
        workspaceId: workspace.id,
        type: ch.type,
        name: ch.name,
        config: { botToken: config.botToken },
      })
      warnings.push(`Channel "${ch.name}" imported as disabled — enable it manually after testing`)
    } catch {
      warnings.push(`Failed to import channel "${ch.name}"`)
    }
  }

  // Ensure always-on capabilities
  await capabilityService.ensureAlwaysOnCapabilities()

  return c.json({
    success: true,
    data: {
      workspace,
      skippedCapabilities,
      warnings,
      modelConfig: parsed.modelConfig,
    },
  })
})

app.get('/:id', async (c) => {
  const { id } = c.req.param()
  const workspace = await workspaceService.findById(id)
  if (!workspace) {
    return c.json({ success: false, error: 'Workspace not found' }, 404)
  }
  return c.json({ success: true, data: workspace })
})

app.get('/:id/settings', async (c) => {
  const { id } = c.req.param()
  const workspace = await workspaceService.findById(id)
  if (!workspace) {
    return c.json({ success: false, error: 'Workspace not found' }, 404)
  }

  return c.json({
    success: true,
    data: {
      id: workspace.id,
      color: workspace.color,
      settings: workspace.settings,
      autoExecute: workspace.autoExecute,
      permissions: workspace.permissions,
    },
  })
})

app.patch('/:id/settings', async (c) => {
  const { id } = c.req.param()
  const body = await c.req.json()
  const parsed = validateBody(updateWorkspaceSettingsSchema, body)
  const workspace = await workspaceService.update(id, parsed)

  return c.json({
    success: true,
    data: {
      id: workspace.id,
      color: workspace.color,
      settings: workspace.settings,
      autoExecute: workspace.autoExecute,
      permissions: workspace.permissions,
    },
  })
})

app.patch('/:id', async (c) => {
  const { id } = c.req.param()
  const body = await c.req.json()
  const parsed = validateBody(updateWorkspaceSchema, body)
  const workspace = await workspaceService.update(id, parsed)
  return c.json({ success: true, data: workspace })
})

app.delete('/:id', async (c) => {
  const { id } = c.req.param()
  try {
    await sandboxService.stopWorkspaceContainer(id)
  } catch (err) {
    logger.warn(`[Workspaces] Failed to stop container for workspace ${id}`, {
      workspaceId: id,
      error: String(err),
    })
  }
  await workspaceService.delete(id)
  return c.json({ success: true, data: { id } })
})

// ── Workspace Export ──────────────────────────────────────

app.get('/:id/export', async (c) => {
  const { id } = c.req.param()
  const workspace = await workspaceService.findById(id)
  if (!workspace) {
    return c.json({ success: false, error: 'Workspace not found' }, 404)
  }

  // Enabled capabilities with decrypted configs
  const enabledCaps = (await capabilityService.getEnabledCapabilitiesForWorkspace(
    id,
  )) as ExportCapabilityRow[]
  const capabilities = enabledCaps.map((cap: ExportCapabilityRow) => {
    const schema = cap.configSchema as
      | import('../capabilities/types.js').ConfigFieldDefinition[]
      | null
    const rawConfig = cap.config as Record<string, unknown> | null
    const config = schema?.length && rawConfig ? decryptConfigFields(schema, rawConfig) : rawConfig
    return { slug: cap.slug, enabled: true, config }
  })

  // Channels with decrypted tokens
  const rawChannels = (await prisma.channel.findMany({
    where: { workspaceId: id },
  })) as ExportChannelRow[]
  const channels = rawChannels.map((ch: ExportChannelRow) => {
    const config = ch.config as Record<string, string>
    let decryptedConfig: Record<string, unknown> = { ...config }
    if (config.botToken) {
      try {
        decryptedConfig.botToken = decrypt(config.botToken)
      } catch {
        /* may not be encrypted */
      }
    }
    return { type: ch.type, name: ch.name, enabled: ch.enabled, config: decryptedConfig }
  })

  // Global model config
  const settings = await settingsService.get()
  const modelConfig = {
    aiProvider: settings.aiProvider,
    aiModel: settings.aiModel,
    roleProviders: buildResolvedRoleProviders(settings),
    mediumModel: settings.mediumModel,
    lightModel: settings.lightModel,
    exploreModel: settings.exploreModel,
    executeModel: settings.executeModel,
    titleModel: settings.titleModel,
    compactModel: settings.compactModel,
    advancedModelConfig: settings.advancedModelConfig,
    embeddingProvider: settings.embeddingProvider,
    embeddingModel: settings.embeddingModel,
    localBaseUrl: settings.localBaseUrl,
    contextLimitTokens: settings.contextLimitTokens,
    maxAgentIterations: settings.maxAgentIterations,
    subAgentExploreMaxIterations: settings.subAgentExploreMaxIterations,
    subAgentAnalyzeMaxIterations: settings.subAgentAnalyzeMaxIterations,
    subAgentExecuteMaxIterations: settings.subAgentExecuteMaxIterations,
    timezone: settings.timezone,
  }

  // Token usage summary
  const usageRows = await prisma.tokenUsage.findMany({
    select: {
      provider: true,
      model: true,
      inputTokens: true,
      outputTokens: true,
    },
  })
  const totals = await prisma.tokenUsage.aggregate({
    _sum: { inputTokens: true, outputTokens: true },
  })
  const usageByModel = new Map<
    string,
    { provider: string; model: string; inputTokens: number; outputTokens: number }
  >()
  for (const row of usageRows) {
    const key = `${row.provider}:${row.model}`
    const existing = usageByModel.get(key)
    if (existing) {
      existing.inputTokens += row.inputTokens
      existing.outputTokens += row.outputTokens
      continue
    }

    usageByModel.set(key, {
      provider: row.provider,
      model: row.model,
      inputTokens: row.inputTokens,
      outputTokens: row.outputTokens,
    })
  }

  const tokenUsage = {
    totalInputTokens: totals._sum.inputTokens ?? 0,
    totalOutputTokens: totals._sum.outputTokens ?? 0,
    byModel: Array.from(usageByModel.values()),
  }

  const exportData: WorkspaceExport = {
    version: 1,
    exportedAt: new Date().toISOString(),
    workspace: {
      name: workspace.name,
      description: workspace.description,
      color: workspace.color,
      autoExecute: workspace.autoExecute,
      settings: workspace.settings as Record<string, unknown> | null,
      permissions: workspace.permissions as { allow: string[] } | null,
    },
    capabilities,
    channels,
    modelConfig,
    tokenUsage,
  }

  const filename = `workspace-${workspace.name.replace(/[^a-zA-Z0-9-_]/g, '_')}-export.json`
  c.header('Content-Disposition', `attachment; filename="${filename}"`)
  c.header('Content-Type', 'application/json')
  return c.json({ success: true, data: exportData })
})

// ── Workspace Capability Overrides ─────────────────────────

app.get('/:id/capabilities', async (c) => {
  const { id } = c.req.param()
  const capabilities = await capabilityService.getWorkspaceCapabilitySettings(id)
  return c.json({ success: true, data: capabilities })
})

app.put('/:id/capabilities/:capabilitySlug', async (c) => {
  const { id, capabilitySlug } = c.req.param()
  const body = await c.req.json()
  if (body.enabled) {
    await capabilityService.enableCapability(id, capabilitySlug, body.config)
  } else {
    await capabilityService.disableCapabilityBySlug(id, capabilitySlug)
  }
  return c.json({ success: true })
})

app.delete('/:id/capabilities/:capabilityId', async (c) => {
  const { id, capabilityId } = c.req.param()
  try {
    await capabilityService.removeCapabilityOverride(id, capabilityId)
    return c.json({ success: true })
  } catch {
    return c.json({ success: false, error: 'Override not found' }, 404)
  }
})

// ── Workspace Container Management ─────────────────────────

app.get('/:id/container/status', async (c) => {
  const { id } = c.req.param()
  const status = await sandboxService.getWorkspaceContainerStatus(id)
  return c.json({ success: true, data: status })
})

app.post('/:id/container/start', async (c) => {
  const { id } = c.req.param()
  const containerId = await sandboxService.startWorkspaceContainerWithCapabilities(id)
  return c.json({ success: true, data: { containerId, status: 'running' } })
})

app.post('/:id/container/stop', async (c) => {
  const { id } = c.req.param()
  await sandboxService.stopWorkspaceContainer(id)
  return c.json({ success: true, data: { status: 'stopped' } })
})

export default app

File: backend/src/clawbuddy/routers/workspaces.py
"""Workspace routes.

Replaces: apps/api/src/routes/workspaces.ts
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, status
from loguru import logger
from sqlalchemy.exc import NoResultFound
from sqlalchemy.ext.asyncio import AsyncSession

from clawbuddy.db.session import get_db
from clawbuddy.lib.responses import fail, ok
from clawbuddy.schemas.workspace import CreateWorkspaceInput, UpdateWorkspaceInput
from clawbuddy.schemas.workspace_export import WorkspaceExport
from clawbuddy.services.workspace import workspace_service

router = APIRouter(tags=["Workspaces"])


@router.get("")
async def list_workspaces(db: AsyncSession = Depends(get_db)) -> dict[str, Any]:
    workspaces = await workspace_service.list(db)
    return ok(workspace_service.serialize_many(workspaces))


@router.post("", status_code=status.HTTP_201_CREATED)
async def create_workspace(
    body: CreateWorkspaceInput,
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    workspace = await workspace_service.create(
        db,
        {
            "name": body.name,
            "description": body.description,
            "color": body.color,
            "settings": body.settings,
        },
    )
    return ok(workspace_service.serialize(workspace))


@router.post("/import", status_code=status.HTTP_201_CREATED)
async def import_workspace(
    body: WorkspaceExport,
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    """Import a workspace from an export file."""
    from clawbuddy.services.capability import capability_service

    skipped_capabilities: list[str] = []
    warnings: list[str] = []

    # Create workspace
    workspace = await workspace_service.create(
        db,
        {
            "name": f"{body.workspace.name} (imported)",
            "description": body.workspace.description,
            "color": body.workspace.color,
            "settings": body.workspace.settings,
        },
    )

    # Apply permissions and autoExecute
    update_data: dict[str, Any] = {"autoExecute": body.workspace.auto_execute}
    if body.workspace.permissions:
        update_data["permissions"] = body.workspace.permissions
    await workspace_service.update(db, workspace.id, update_data)

    # Enable capabilities
    for cap in body.capabilities:
        if not cap.enabled:
            continue
        try:
            await capability_service.enable_capability(
                db, workspace.id, cap.slug, cap.config
            )
        except Exception:
            skipped_capabilities.append(cap.slug)

    # Create channels (disabled by default)
    from clawbuddy.services.channel import channel_service

    for ch in body.channels:
        try:
            config = ch.config
            if not config.get("botToken"):
                warnings.append(f'Channel "{ch.name}" skipped — no bot token')
                continue
            await channel_service.create(
                db,
                workspace_id=workspace.id,
                channel_type=ch.type,
                name=ch.name,
                config={"botToken": config["botToken"]},
            )
            warnings.append(
                f'Channel "{ch.name}" imported as disabled — enable it manually after testing'
            )
        except Exception:
            warnings.append(f'Failed to import channel "{ch.name}"')

    # Ensure always-on capabilities
    await capability_service.ensure_always_on_capabilities(db)

    return ok(
        {
            "workspace": workspace_service.serialize(workspace),
            "skippedCapabilities": skipped_capabilities,
            "warnings": warnings,
            "modelConfig": body.model_config_data.model_dump(by_alias=True),
        }
    )


@router.get("/{workspace_id}")
async def get_workspace(
    workspace_id: str,
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    workspace = await workspace_service.find_by_id(db, workspace_id)
    if not workspace:
        return fail("Workspace not found", status_code=404)
    return ok(workspace_service.serialize(workspace))


@router.patch("/{workspace_id}")
async def update_workspace(
    workspace_id: str,
    body: UpdateWorkspaceInput,
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    data: dict[str, Any] = {}
    if body.name is not None:
        data["name"] = body.name
    if body.description is not None:
        data["description"] = body.description
    if body.color is not None:
        data["color"] = body.color
    if body.permissions is not None:
        data["permissions"] = body.permissions.model_dump()
    if body.settings is not None:
        data["settings"] = body.settings
    if body.auto_execute is not None:
        data["autoExecute"] = body.auto_execute
    workspace = await workspace_service.update(db, workspace_id, data)
    return ok(workspace_service.serialize(workspace))


@router.delete("/{workspace_id}")
async def delete_workspace(
    workspace_id: str,
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    try:
        from clawbuddy.services.sandbox import sandbox_service
        await sandbox_service.stop_workspace_container(workspace_id)
    except Exception as err:
        logger.warning(f"[Workspaces] Failed to stop container for workspace {workspace_id}: {err}")
    await workspace_service.delete(db, workspace_id)
    return ok({"id": workspace_id})


# ── Workspace Export ──────────────────────────────────────

@router.get("/{workspace_id}/export")
async def export_workspace(
    workspace_id: str,
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    from clawbuddy.db.models import Channel, TokenUsage
    from clawbuddy.lib.llm_resolver import build_resolved_role_providers
    from clawbuddy.services.capability import capability_service
    from clawbuddy.services.config_validation import decrypt_config_fields
    from clawbuddy.services.crypto import decrypt
    from clawbuddy.services.settings_service import settings_service
    from sqlalchemy import func, select

    workspace = await workspace_service.find_by_id(db, workspace_id)
    if not workspace:
        return fail("Workspace not found", status_code=404)

    # Enabled capabilities with decrypted configs
    enabled_caps = await capability_service.get_enabled_capabilities_for_workspace(
        db, workspace_id
    )
    capabilities = []
    for cap in enabled_caps:
        schema = cap.get("configSchema")
        raw_config = cap.get("config")
        config = (
            decrypt_config_fields(schema, raw_config)
            if schema and raw_config
            else raw_config
        )
        capabilities.append({"slug": cap["slug"], "enabled": True, "config": config})

    # Channels with decrypted tokens
    ch_result = await db.execute(
        select(Channel).where(Channel.workspace_id == workspace_id)
    )
    raw_channels = ch_result.scalars().all()
    channels = []
    for ch in raw_channels:
        config = dict(ch.config or {})
        if config.get("botToken"):
            try:
                config["botToken"] = decrypt(config["botToken"])
            except Exception:
                pass
        channels.append(
            {"type": ch.type, "name": ch.name, "enabled": ch.enabled, "config": config}
        )

    # Global model config
    settings = await settings_service.get(db)
    model_config = {
        "aiProvider": settings["aiProvider"],
        "aiModel": settings["aiModel"],
        "roleProviders": build_resolved_role_providers(settings),
        "mediumModel": settings.get("mediumModel"),
        "lightModel": settings.get("lightModel"),
        "exploreModel": settings.get("exploreModel"),
        "executeModel": settings.get("executeModel"),
        "titleModel": settings.get("titleModel"),
        "compactModel": settings.get("compactModel"),
        "advancedModelConfig": settings.get("advancedModelConfig"),
        "embeddingProvider": settings["embeddingProvider"],
        "embeddingModel": settings.get("embeddingModel"),
        "localBaseUrl": settings.get("localBaseUrl"),
        "contextLimitTokens": settings.get("contextLimitTokens"),
        "maxAgentIterations": settings.get("maxAgentIterations"),
        "subAgentExploreMaxIterations": settings.get("subAgentExploreMaxIterations"),
        "subAgentAnalyzeMaxIterations": settings.get("subAgentAnalyzeMaxIterations"),
        "subAgentExecuteMaxIterations": settings.get("subAgentExecuteMaxIterations"),
        "timezone": settings.get("timezone"),
    }

    # Token usage summary
    usage_result = await db.execute(
        select(
            TokenUsage.provider,
            TokenUsage.model,
            func.sum(TokenUsage.input_tokens).label("input_sum"),
            func.sum(TokenUsage.output_tokens).label("output_sum"),
        ).group_by(TokenUsage.provider, TokenUsage.model)
    )
    usage_rows = usage_result.all()

    totals_result = await db.execute(
        select(
            func.sum(TokenUsage.input_tokens).label("input_sum"),
            func.sum(TokenUsage.output_tokens).label("output_sum"),
        )
    )
    totals = totals_result.one()

    token_usage = {
        "totalInputTokens": totals.input_sum or 0,
        "totalOutputTokens": totals.output_sum or 0,
        "byModel": [
            {
                "provider": row.provider,
                "model": row.model,
                "inputTokens": row.input_sum or 0,
                "outputTokens": row.output_sum or 0,
            }
            for row in usage_rows
        ],
    }

    from datetime import datetime, timezone

    export_data = {
        "version": 1,
        "exportedAt": datetime.now(timezone.utc).isoformat(),
        "workspace": {
            "name": workspace.name,
            "description": workspace.description,
            "color": workspace.color,
            "autoExecute": workspace.auto_execute,
            "settings": workspace.settings,
            "permissions": workspace.permissions,
        },
        "capabilities": capabilities,
        "channels": channels,
        "modelConfig": model_config,
        "tokenUsage": token_usage,
    }

    return ok(export_data)


# ── Workspace Capability Overrides ─────────────────────────

@router.get("/{workspace_id}/capabilities")
async def get_workspace_capabilities(
    workspace_id: str,
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    from clawbuddy.services.capability import capability_service

    capabilities = await capability_service.get_workspace_capability_settings(
        db, workspace_id
    )
    return ok(capabilities)


@router.put("/{workspace_id}/capabilities/{capability_slug}")
async def update_workspace_capability(
    workspace_id: str,
    capability_slug: str,
    body: dict[str, Any],
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    from clawbuddy.services.capability import capability_service

    if body.get("enabled"):
        await capability_service.enable_capability(
            db, workspace_id, capability_slug, body.get("config")
        )
    else:
        await capability_service.disable_capability_by_slug(
            db, workspace_id, capability_slug
        )
    return ok(None)


@router.delete("/{workspace_id}/capabilities/{capability_id}")
async def remove_capability_override(
    workspace_id: str,
    capability_id: str,
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    from clawbuddy.services.capability import capability_service

    try:
        await capability_service.remove_capability_override(
            db, workspace_id, capability_id
        )
        return ok(None)
    except Exception:
        return fail("Override not found", status_code=404)


# ── Workspace Container Management ─────────────────────────

@router.get("/{workspace_id}/container/status")
async def get_container_status(workspace_id: str) -> dict[str, Any]:
    from clawbuddy.services.sandbox import sandbox_service

    try:
        container_status = await sandbox_service.get_workspace_container_status(
            workspace_id
        )
    except NoResultFound:
        return fail("Workspace not found", status_code=404)
    return ok(container_status)


@router.post("/{workspace_id}/container/start")
async def start_container(workspace_id: str) -> dict[str, Any]:
    from clawbuddy.services.sandbox import sandbox_service

    try:
        container_id = (
            await sandbox_service.start_workspace_container_with_capabilities(
                workspace_id
            )
        )
    except NoResultFound:
        return fail("Workspace not found", status_code=404)
    return ok({"containerId": container_id, "status": "running"})


@router.post("/{workspace_id}/container/stop")
async def stop_container(workspace_id: str) -> dict[str, Any]:
    from clawbuddy.services.sandbox import sandbox_service

    try:
        await sandbox_service.stop_workspace_container(workspace_id)
    except NoResultFound:
        return fail("Workspace not found", status_code=404)
    return ok({"status": "stopped"})
